### DISCUSSION TITLE
Принцип подбора треков craftTracks

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#80] [Принцип работы]
С удовольствием пользуюсь этой функцией, но не очень понимаю принципа подбора треков в ней. Описание функции несколько суховато. 

Например, что значит "полученный для каждой пятерки элементов"?
Берется первые пять треков, из этой пятерки берется первый трек и к нему подбирается похожий, потом так же происходит со следующими треками этой пятерки, и таким образом получается пятерка похожих треков, которая отдается в результат. Потом берется следующая пятерка и цикл повторяется. Зачем тогда нужна такая пятёрка, если это можно делать просто по одному треку?
Или пятерка похожих треков каким-то образом формируется сразу из общей суммы полученных пяти треков. Каким?

И по какому принципу происходит отбор "похожести" трека? Это какой-то алгоритм самого Spotify? Он известен? Или это логика Goofy? Особенно интересно как это происходит по умолчанию, когда в аргументах только массив треков.

Какие треки берутся из массива - первые, или случайные? Например если постоянно использовать функцию, для получения похожих треков от плейлиста "Любимые треки" и не указывать размер, то для того чтобы каждый раз результат был разным, перед этим нужно выбрать из него 100 случайных треков. Или это не обязательно?

**Comment by Chimildic** [comment]
> принципу происходит отбор "похожести" трека?

Рекомендации Spotify через [getRecomTracks](https://chimildic.github.io/goofy/#/func?id=getrecomtracks).

> что значит "полученный для каждой пятерки элементов"?

Если прочитать описание к `getRecomTracks`, можно заметить `seed_* - до 5 значений`:
- `seed` в буквальном переводе является `зерном`. `id` наших треков становятся зернами для рекомендаций. Spotify своей базой и алгоритмами _выращивает_ на них абстрактное поле и возвращает нам. Можно предположить, что принципом отбора являются социальные связи (один пользователь слушает трек A и B, а второй только А - значит второму можно порекомендовать B), а также близость [по характеристикам](https://chimildic.github.io/goofy/#/guide?id=Особенности-трека-features) (`energy`, `tempo` и пр.)
- Ограничение в 5 элементов диктует Spotify API
> A comma separated list of Spotify IDs for seed *. **Up to 5 seed values may be provided** in any combination of seed_artists, seed_tracks and seed_genres.

> Какие треки берутся из массива - первые, или случайные?

Таким образом, `craftTracks` разбивает массив треков на группы по 5 элементов (подряд, ничего не пропускает). Если не хватает, последняя группа может иметь `id` одного трека, например. И просит у Spotify для каждой группы рекомендации. Затем объединяет все в один массив и удаляет дубликаты. 

В описании к `craftTracks` есть примечание

> Можно частично повлиять на формируемые пятерки элементов. До вызова функции применив одну из сортировок Order.

Что это значит? Изменяется сортировка массива. `craftTracks` проходя подряд по трекам сформирует уже другие пятерки-группы. Отсюда Spotify получит другой набор _зерн_ и, скорее всего, вернет уже другие рекомендации. 
Аналогичный результат получится если применить `Selector.keepRandom` к исходному массиву.

> Особенно интересно как это происходит по умолчанию, когда в аргументах только массив треков.

Указав только массив треков, по умолчанию подставится ключ `seed_tracks`. Остальная логика не изменяется. 

Поскольку это `getRecomTracks`, можно указывать дополнительные параметры. Они помещаются по ключу `query`. То есть эти параметры будут отправляться с каждой пятеркой-группой.
```js
query: {
      limit: 20, // по умолчанию и максимум 100
      min_energy: 0.4,
      min_popularity: 60,
      // target_popularity: 60,
}
```

Кстати, рекомендую попробовать [Filter.replaceWithSimilar](https://chimildic.github.io/goofy/#/func?id=replacewithsimilar). Он отправляет запросы `getRecomTracks` и в качестве параметров использует `target_*` ключи со значениями исходных треков, что в теории дает рекомендации наиболее близкие по характеристикам трека. 
```js
// часть кода replaceWithSimilar
Source.createUrlForRecomTracks({
    seed_tracks: t.id,
    seed_artists: t.artists[0].id,
    target_danceability: features[t.id].danceability,
    target_energy: features[t.id].energy,
    target_acousticness: features[t.id].acousticness,
    target_valence: features[t.id].valence,
})
```

**Reply by Bedmit** [reply]
Спасибо. По craftTracks более-менее понял. Зато теперь вопрос по Filter.replaceWithSimilar. ))
Я правильно понимаю, что при значениях 
`Filter.replaceWithSimilar(tracks, RecentTracks.get(), Source.getSavedTracks());`
похожие треки будут не из числа залайканых и ранее прослушанных?

**Reply by Chimildic** [reply]
Если в `tracks` есть трек из лайков или истории - он станет зерном рекомендаций. Из них выберится случайный трек и встанет на место исходного трека.

**Reply by Bedmit** [reply]
Поиграл еще не много с `replaceWithSimilar` и заметил одну особенность. Если для трека невозможно подобрать рекомендации, я так понимаю из-за очень маленького количества прослушиваний, то функция не заменяет этот трек. Даже если этот трек есть и в лайках, и в истории. Мне кажется логичней если он в таком случае будет хотя бы просто удаляться. Т.е. в результатах обработки плейлиста в любом случае не должно оставаться залайканых и ранее прослушанных треков.

**Reply by Chimildic** [reply]
@Bedmit возможно дело в том, что код не обновлен. Поддержку сразу нескольких массивов на замену только недавно добавил. Но да, если трека на замену нет, оставался прежний. 
Обновил, чтобы он удалялся. Заодно добавил пару параметров (изначально не все использовались). Может поточнее выйдут рекомендации.

**Reply by Bedmit** [reply]
Да, после обновления все ок. На счет улучшения точности рекомендаций пока сложно сказать. Но в любом случае для получения похожих треков на основе именно массива треков, кмк эта функция подходит больше. Она не ломает исходный состав в пользу более популярных треков, как это делает `craftTracks`. Зато последняя отлично подходит для радио по одному треку.

Еще хотел кое-что уточнить. Почему в недавнем примере с фильтром `filterByDeviation` один из аргументов имеет вид `Selector.sliceCopy(savedTracks)` вместо `Source.getSavedTracks()`? Я правильно понимаю, что это для того чтобы не тратить ресурсы на повторное получение уже ранее полученного массива треков? Т.е. мне в используемых мной скриптах хорошо бы тоже так исправить аргументы. И `RecentTracks.get()` тоже поменять на `Selector.sliceCopy(History)` раз уж у меня есть `let History = RecentTracks.get();`

И что значит может принять сразу несколько массивов для замены? Ограничений на количество нет вообще?

**Reply by Chimildic** [reply]
> один из аргументов имеет вид Selector.sliceCopy(savedTracks)

Да, чтобы не делать дополнительные запросы тех же треков. Это полезно, когда другие функции влияют на содержание. Фильтр какой нибудь, например. В данном случае это похоже "остаточный" код. Сейчас можно просто `savedTracks`, потому что фильтр не изменит их содержание, насколько я вижу. В случае `RecentTracks.get()` эффект не так значим, потому что файл читаем, а не запросы отправляем. Может чуть быстрее станет, но не значительно.

> принять сразу несколько массивов для замены?

Как в `push`, второй и последующие массивы при вызове `replaceWithSimilar` объединятся в один
```js
Filter.replaceWithSimilar(tracks, savedTracks, RecentTracks.get(3500));
```

**Reply by Chimildic** [reply]
Последняя правка с удалением трека в `replaceWithSimilar` оказалась неудачной. Удаляло и те треки, для которых замена не требовалась. 
Мне не удается поймать трек, для которого не дают рекомендации, чтобы корректно условие поставить.

**Reply by Bedmit** [reply]
Угу, я видел в соседней теме.
Правда при моем сценарии использования это было даже предпочтительней. Мне не нужны в результатах треки массива-источника. Для меня вообще был бы более интересен вариант функции, когда абсолютно все треки плейлиста заменяются на похожие. Чтобы в результатах не было залайканых и ранее прослушанных треков, а так же треков плейлиста-источника. Что-то типа 
  ```
let tracks = Source.getPlaylistTracks('Название', 'ID');
 Filter.replaceWithSimilar(tracks);
```
где все это делается по умолчанию.
Сейчас для этих целей вторая строка 
`Filter.replaceWithSimilar(tracks, tracks, Source.getSavedTracks(), RecentTracks.get()); `
